app/models/user.rb

```diff
@@ -203,16 +206,23 @@ def self.upload_user_list(user_class, user_list, encoding)
     user_columns.push(:time_zone)
     users.each { |u| u.push("#{u[first_name_i]} #{u[last_name_i]}") }
     users.each { |u| u.push(Time.zone.name) }
-
-    existing_user_ids = user_class.all.pluck(:id)
-    imported = nil
     parsed[:invalid_records] = ''
+    return parsed if users.empty?
+    existing_user_ids = user_class.all.pluck(:id)
+    imported_ids = []
+    successful_imports = []
+    all_user_names = []
+
     User.transaction do
-      imported = user_class.import user_columns, users, on_duplicate_key_update: {
-        conflict_target: [:user_name],
-        columns: [:last_name, :first_name, :section_id, :email, :id_number, :display_name, :time_zone]
-      }
-      User.where(id: imported.ids).each do |user|
+      user_hash = users.collect { |record| Hash[user_columns.zip record] }
+      user_hash.each do |user|
+        user[:type] = user_class.name
+        all_user_names.push(user[:user_name])
+      end
+      imported = user_class.upsert_all(user_hash, unique_by: :user_name, returning: %w[id user_name])
+      successful_imports = imported.rows.map { |x| [x[1]] }.flatten
```

Comment:
If you just use `x[1]` instead of wrapping it in a list, you can remove the `flatten`. Same as the line below as well.